File: 210-course-schedule-ii/210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        inDegrees = [0] * numCourses
        graph = defaultdict(list)
        
        for course, pre in prerequisites:
            graph[pre].append(course)
            inDegrees[course] += 1
            
        q, result = deque(), []
        
        for i in range(numCourses):
            if inDegrees[i] == 0:
                q.append(i)
        
        while q:
            course = q.popleft()
            result.append(course)
            for neighbor in graph[course]:
                inDegrees[neighbor] -= 1
                if inDegrees[neighbor] == 0:
                    q.append(neighbor)
                    
        return result if len(result) == numCourses else []

File: 210-course-schedule-ii/210-course-schedule-ii.cpp
class Solution {
public:
    
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        int n = numCourses;
        vector<int> ans, indeg(n);
        map<int, set<int>> adj;
        
        for (auto pr: prerequisites) {
            adj[pr[1]].insert(pr[0]);
            indeg[pr[0]]++;
        }
        queue<int> q;
        for (int i = 0; i < n; i++)
            if (indeg[i] == 0)
                q.push(i);
        
        while (!q.empty()) {
            int u = q.front();
            ans.push_back(u);
            q.pop();
            for (auto v: adj[u])
                if (--indeg[v] == 0)
                    q.push(v);
        }
        if (ans.size() == n)
            return ans;
        else
            return {};
    }
};